premailer/cache.py

```diff
@@ -0,0 +1,73 @@
+import functools
+
+class _HashedSeq(list):
+    # # From CPython
+    __slots__ = 'hashvalue'
+
+    def __init__(self, tup, hash=hash):
+        self[:] = tup
+        self.hashvalue = hash(tup)
+
+    def __hash__(self):
+        return self.hashvalue
+
+## if we only have nonlocal
+class _Cache(object):
+    def __init__(self):
+        self.off = False
+        self.missed = 0
+        self.cache = {}
+
+def function_cache(expected_max_entries=1000):
```

Comment:
Excellent. Yes I was worried about copyright. The whole project is Python which protects us but since premailer is used inside corporate projects it's better to be safe than sorry.